discovery_engine_core/core/src/engine.rs

```diff
@@ -140,17 +141,23 @@ where
         })
     }
 
-    /// Serializes the state of the `Engine`.
-    ///
-    /// The result can be used with [`Engine::new`] to restore it.
+    /// Serializes the `Engine` and `Ranker` state.
     pub fn serialize(&self) -> Result<Vec<u8>, Error> {
         let stacks_data: HashMap<&StackId, &StackData> = self
             .stacks
             .iter()
             .map(|(id, stack)| (id, &stack.data))
             .collect();
 
-        bincode::serialize(&stacks_data).map_err(Error::Serialization)
+        let engine_state = bincode::serialize(&stacks_data).map_err(Error::Serialization)?;
+        let ranker_state = self.ranker.serialize()?;
```

Comment:
should we map the error? otherwise we would get an `Error::Ranker()` instead of an `Error::Serialization()` which might be misleading. this would require some changes to the error variants though.
```suggestion
        let ranker_state = self.ranker.serialize().map_err(Error::Serialization)?;
```

Reply:
Unfortunately, this doesn't work because `self.ranker.serialize()` returns a `GenericError` and `Error::Serialization` expects a `bincode::Error`. We could introduce a separate `RankerSerialization` error or change the source of `Error::Serialization` to `GenericError`

Reply:
for now it's probably fastest to change the source of the error, this is what i had in mind with

> this would require some changes to the error variants though.

Reply:
done